tests/contextwatchdog.js

```diff
@@ -0,0 +1,588 @@
+
+describe( 'ContextWatchdog', () => {
+	let element1, element2;
+	let watchdog;
+	let originalErrorHandler;
+
+	beforeEach( () => {
+		element1 = document.createElement( 'div' );
+		element2 = document.createElement( 'div' );
+
+		document.body.appendChild( element1 );
+		document.body.appendChild( element2 );
+
+		originalErrorHandler = window.onerror;
+		window.onerror = sinon.spy();
+	} );
+
+	afterEach( () => {
+		window.onerror = originalErrorHandler;
+
+		element1.remove();
+		element2.remove();
+
+		sinon.restore();
+	} );
+
+	it( 'should disable adding items once the Watchdog is destroyed', async () => {
+		watchdog = ContextWatchdog.for( Context, {} );
+
+		await watchdog.destroy();
+
+		let err;
+
+		try {
+			await watchdog.add( {
+				editor2: {
+					type: 'editor',
+					creator: ( el, config ) => ClassicTestEditor.create( el, config ),
+					sourceElementOrData: element1,
+					config: {}
+				},
+			} );
+		} catch ( _err ) {
+			err = _err;
+		}
+
+		expect( err ).to.be.instanceOf( Error );
+		expect( err.message ).to.match( /Cannot add items to destroyed watchdog\./ );
+	} );
+
+	describe( 'for scenario with no items', () => {
+		it( 'should create only context', async () => {
+			watchdog = ContextWatchdog.for( Context, {} );
+
+			await watchdog.waitForReady();
+
+			expect( watchdog.context ).to.be.instanceOf( Context );
+
+			await watchdog.destroy();
+		} );
+
+		it( 'should have proper states', async () => {
+			watchdog = ContextWatchdog.for( Context, {} );
+
+			expect( watchdog.state ).to.equal( 'initializing' );
+
+			await watchdog.waitForReady();
+
+			expect( watchdog.state ).to.equal( 'ready' );
+
+			await watchdog.destroy();
+
+			expect( watchdog.state ).to.equal( 'destroyed' );
+		} );
+
+		it( 'should set custom destructor if provided', async () => {
+			const mainWatchdog = new ContextWatchdog();
+			const customDestructor = sinon.spy( context => context.destroy() );
+
+			mainWatchdog.setCreator( config => Context.create( config ) );
+			mainWatchdog.setDestructor( customDestructor );
+			mainWatchdog.create();
+
+			await mainWatchdog.destroy();
+
+			sinon.assert.calledOnce( customDestructor );
+		} );
+
+		it( 'should log if an error happens during the component destructing', async () => {
+			const mainWatchdog = new ContextWatchdog();
+
+			const consoleErrorStub = sinon.stub( console, 'error' );
+			const err = new Error( 'foo' );
+			mainWatchdog.setCreator( config => Context.create( config ) );
+			mainWatchdog.setDestructor( async editor => {
+				await editor.destroy();
+
+				throw err;
+			} );
+
+			await mainWatchdog.create();
+
+			await mainWatchdog._restart();
+
+			sinon.assert.calledWith(
+				consoleErrorStub,
+				'An error happened during destructing.',
+				err
+			);
+
+			mainWatchdog.setDestructor( editor => editor.destroy() );
+
+			await mainWatchdog.destroy();
+		} );
+
+		it( 'should handle the Watchdog configuration', async () => {
+			// TODO
```

Comment:
:thinking:

Reply:
I've added the missing test :sweat\_smile: